### PR TITLE
feat(pose_estimator_arbiter): add pose_estimator_arbiter to realize multi localizer

### PR DESCRIPTION
**Pull request by KYabuuchi**
## Description

This PR implement a mechanism to launch multiple pose_estimators, and allow for pausing and resuming specific ones according to the situation.

[The package README](https://github.com/KYabuuchi/autoware.universe/tree/feat/multi_localizer/localization/pose_estimator_arbiter) describes the purpose in detail.

## Related links

[[Proposal] Runtime Localization Module Switching for Multi-Pose Estimators #3878](https://github.com/orgs/autowarefoundation/discussions/3878)

## Tests performed

I have confirmed that ndt/yabloc/eagleye/artag is available as before with `logging_simulator.launch.xml` .
Also, I confirmed that by specifying multiple pose_estimators at startup, pose_estimator_arbiter is launched and arbitrates the pose_estimators according to the situation.

### sample data
[sample data (rosbag & map)](https://drive.google.com/file/d/1MxLo1Sw6PdvfkyOYf_9A5dZ9uli1vPvS/view)
The rosbag is simulated data created by [AWSIM](https://tier4.github.io/AWSIM/). The map is an edited version of the original map data published on the AWSIM documentation page to make it suitable for multiple pose_estimators.

### how to launch 

When start a single pose_estimator:
```shell
ros2 launch autoware_launch logging_simulator.launch.xml \
    map_path:=$HOME/MAP vehicle_model:=sample_vehicle sensor_model:=awsim_sensor_kit \
    pose_source:=ndt # or yabloc, eagleye or artag
```

When start multiple pose_estimator:
```bash
ros2 launch autoware_launch logging_simulator.launch.xml \
    map_path:=$HOME/MAP vehicle_model:=sample_vehicle sensor_model:=awsim_sensor_kit \
    pose_source:=ndt_yabloc_eagleye_artag # or ndt_yabloc, eagleye_artag and so on
```

### expected behavior
When start a single pose_estimator, it launches the same nodes as before.

When start multiple pose_estimator, it works as follows.

https://github.com/autowarefoundation/autoware.universe/assets/24854875/62b26fdd-dcf0-4b1c-a1a0-ecd633413196

> [!IMPORTANT]
> We can visualize the debug prints of pose_estimator_arbiter at any time with the following command
> (I recommend enabling logging, because with the default rviz settings, we do not know which pose_estimator is running.)
```
ros2 service call /localization/pose_estimator_arbiter/config_logger \
    logging_demo/srv/ConfigLogger \
   '{logger_name: localization.pose_estimator_arbiter , level: debug}'
```

> [!IMPORTANT]
> Eagleye requires a GNSS doppler, but AWSIM does not generate it. So I have stored my pseudo-generated dopplers in rosbag. Because of that, Eagleye's estimated position is not accurate.

## Notes for reviewers

> [!TIP]
> If you want to create an rviz screen like the demonstration video, please visualize the following topics.
> `/localization/pose_estimator_arbiter/debug/marker_array`
> `/localization/pose_estiamtor/points_aligned`
> `/localization/pose_estimator/yabloc/image_processing/lanelet2_overlay_image`
> `/localization/pose_estimator/yabloc/pf/predicted_particles_marker`
> `/localization/pose_estimator/ar_tag_based_localizer/debug/image`
> `/localization/eagleye/pose_with_covariance/to_relay`

Sorry for the large number of commits.

## Interface changes

The default behavior remains the same. 
Users can now specify multiple pose_estimators for `pose_source` . By default, ndt is launched.

## Effects on system behavior

The default behavior remains the same. 
Users can now specify multiple pose_estimators for `pose_source` . By default, ndt is launched.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: localization/yabloc/yabloc_particle_filter/src/prediction/predictor.cpp [hunk 1]
```diff
@@ -53,26 +53,54 @@ Predictor::Predictor()
...
   // Optional modules
   if (declare_parameter<bool>("visualize", false)) {
     visualizer_ptr_ = std::make_unique<ParticleVisualizer>(*this);
   }
 }
 
+void Predictor::on_service(
+  SetBool::Request::ConstSharedPtr request, SetBool::Response::SharedPtr response)
+{
+  RCLCPP_INFO_STREAM(get_logger(), "particle filter activatation is set to " << request->data);
```
Comment (SakodaShintaro) [inline localization/yabloc/yabloc_particle_filter/src/prediction/predictor.cpp:88]: There seems to be a typo.

File: localization/pose_estimator_arbiter/test/pose_estimator_arbiter.test.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+        # Publish dummy input topics
+        for _ in range(10):
+            self.publish_input_topics()
+            rclpy.spin_once(self.test_node, timeout_sec=0.1)
+
+        # Wait 0.5 second for all topics to be subscribed
+        self.spin_for(0.5)
+
+        # Confirm that yabloc topic is relayed and ndt toic is not relayed.
```
Comment (SakodaShintaro) [inline localization/pose_estimator_arbiter/test/pose_estimator_arbiter.test.py:195]: There seems to be a typo.

**Review by SakodaShintaro — COMMENTED** [review]
Thank you for the wonderful pull request!
I have some comments, so I would appreciate it if you could answer them.

**Comment by KYabuuchi** [comment]
Hi. All reviewers.

I am going to remove the `pcd_occuancpy`, `artag_position` rules from the main stream and implement rules that switch all pose_estimators by reference to the vector map.
This will simplify the source code and README.

File: launch/tier4_localization_launch/launch/pose_twist_estimator/ar_tag_based_localizer.launch.xml [hunk 3]
```diff
@@ -1,9 +1,13 @@
 <?xml version="1.0"?>
 <launch>
   <group>
+    <arg name="multi_localizer_mode" default="false" description="Set to true when multiple pose_estimators are run simultaneously."/>
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply launch/tier4_localization_launch/launch/pose_twist_estimator/ar_tag_based_localizer.launch.xml]: Thanks for comments.  :+1: 
I moved them to pose_twist_estimator.launch.xml in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/2d66593eff7f05c7a8d43f56e6994177e528ea35

File: launch/tier4_localization_launch/launch/pose_twist_estimator/pose_twist_estimator.launch.xml [hunk 4]
```diff
@@ -1,148 +1,106 @@
...
+      <arg name="yabloc_enabled" value="$(var use_yabloc_pose)"/>
+      <arg name="gnss_enabled" value="true"/>
+      <arg name="ekf_enabled" value="true"/>
+      <arg name="stop_check_enabled" value="$(var stop_check_enabled)"/>
+      <arg name="config_file" value="$(var pose_initializer_param_path)"/>
+      <arg name="sub_gnss_pose_cov" value="$(var sub_gnss_pose_cov)"/>
+    </include>
+
+    <!-- automatic_pose_initializer -->
+    <group if="$(var gnss_enabled)">
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply launch/tier4_localization_launch/launch/pose_twist_estimator/pose_twist_estimator.launch.xml:104]: I defined the `gnss_enabled` at the top of this file. (This does not affect system behavior.)
https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/f0943803207f8e50dc37af9cd333f8f61e1e464a

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace pose_estimator_arbiter
+{
+class PoseEstimatorArbiter : public rclcpp::Node
+{
+public:
+  using SetBool = std_srvs::srv::SetBool;
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter.hpp:39]: They do not need to be written as public. I correctly configured the access level in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/1a0a2574d58b5bef97d6bc216ff9717f773de9f8

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_name.hpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef POSE_ESTIMATOR_ARBITER__POSE_ESTIMATOR_NAME_HPP_
+#define POSE_ESTIMATOR_ARBITER__POSE_ESTIMATOR_NAME_HPP_
+
+namespace pose_estimator_arbiter
+{
+enum class PoseEstimatorName : int { ndt = 1, yabloc = 2, eagleye = 4, artag = 8 };
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_name.hpp]: Thanks. `PoseEstimatorType` makes more sense to me. I renamed it in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/1bd74f7a8b6c3c7769ab9d328893182e1f7aa36f

File: localization/pose_estimator_arbiter/test/pose_estimator_arbiter.test.py (hunk 2 above)
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/test/pose_estimator_arbiter.test.py]: Fixed https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/aae8e8e1c94510c31f2d7dc76d1bb3b5a50c54a2

File: localization/yabloc/yabloc_particle_filter/src/prediction/predictor.cpp (hunk 1 above)
Comment (KYabuuchi) [reply localization/yabloc/yabloc_particle_filter/src/prediction/predictor.cpp]: Fixed https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/aae8e8e1c94510c31f2d7dc76d1bb3b5a50c54a2

File: localization/yabloc/yabloc_image_processing/launch/yabloc_image_processing.launch.xml [hunk 7]
```diff
@@ -1,7 +1,8 @@
 <launch>
   <arg name="src_image"/>
   <arg name="src_info"/>
-  <arg name="resized_image" default="undistorted/image_raw"/>
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/yabloc/yabloc_image_processing/launch/yabloc_image_processing.launch.xml:4]: I am going to make the another PR for yabloc

[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/yabloc/yabloc_image_processing/launch/yabloc_image_processing.launch.xml:4]: I have created the another PR to split this PR https://github.com/autowarefoundation/autoware.universe/pull/6076

If you can please review that too :pray:

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    sub_localization_pose_cov_ = create_subscription<PoseCovStamped>(
+      "~/input/pose_with_covariance", 5, shared_data_->localization_pose_cov.create_callback());
+    sub_point_cloud_map_ = create_subscription<PointCloud2>(
+      "~/input/pointcloud_map", latch_qos, shared_data_->point_cloud_map.create_callback());
+    sub_vector_map_ = create_subscription<HADMapBin>(
+      "~/input/vector_map", latch_qos, shared_data_->vector_map.create_callback());
+    sub_initialization_state_ = create_subscription<InitializationState>(
+      "~/input/initialization_state", latch_qos,
+      shared_data_->initialization_state.create_callback());
+  }
```
Comment (TaikiYamada4) [inline localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp:107]: Is it necessary to subscribe all of these even when Autoware doesn't use all pose_estimators?

For example, `/localization/eagleye/pose_with_covariance/to_relay` appears in `ros2 topic list` even when `pose_source:=ndt_yabloc`, and pose_estimator_arbiter will subscribe `/sensing/camera/traffic_light/image_raw` even when `pose_source:=ndt_eagleye`.

Is it possible to declare them only when the according pose_estimator_name exists?

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  {
+    diag_status.name = "localization: " + std::string(this->get_name());
+    diag_status.hardware_id = this->get_name();
+
+    diag_status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    diag_status.message = "OK";
+
+    diagnostic_msgs::msg::KeyValue key_value_msg;
+    key_value_msg.key = "state";
+    key_value_msg.value = "Diagnostics is not implemented.";
```
Comment (TaikiYamada4) [inline localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp:164]: What does "Diagnostics is not implemented." mean here? (It IS implemented after all).
Should I take this as "Further details to be implemented later."?

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/switch_rule/map_based_rule_core.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+#include "pose_estimator_arbiter/rule_helper/grid_info.hpp"
+#include "pose_estimator_arbiter/switch_rule/map_based_rule.hpp"
+
+namespace pose_estimator_arbiter::switch_rule
+{
+std::unordered_map<PoseEstimatorType, bool> MapBasedRule::update()
+{
+  auto ret = update_impl();
+  RCLCPP_DEBUG(get_logger(), debug_string_.c_str());
```
Comment (TaikiYamada4) [inline localization/pose_estimator_arbiter/src/pose_estimator_arbiter/switch_rule/map_based_rule_core.cpp:23]: I can avoid the compile error 
```
/opt/ros/humble/include/rcutils/rcutils/logging_macros.h:72:18: error: format not a string literal and no format arguments [-Werror=format-security]
```
by changing this row as the following. (adding "%s",)
```
RCLCPP_DEBUG(get_logger(), "%s", debug_string_.c_str());
```

Is there any problem for doing this?

**Review by TaikiYamada4 — COMMENTED** [review]
I have three comments to your PR, so could you please take a look at them.
Plus, I get 
```
/usr/include/pcl-1.12/pcl/impl/point_types.hpp:241:12: warning: ISO C++ prohibits anonymous structs [-Wpedantic]
```
and
```
pose_estimator_arbiter::switch_rule::MapBasedRule::update()':
/opt/ros/humble/include/rcutils/rcutils/logging_macros.h:72:18: error: format not a string literal and no format arguments [-Werror=format-security]
   72 |       rcutils_log(&__rcutils_logging_location, severity, name, __VA_ARGS__);
```
when I compile in the general way.
```
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release
```

I can avoid this by the following command, and checked that pose_estimator_arbiter works as it used to be, but do you have other treatments to deal with "-Wno-error=pedantic"? 
```
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-Wno-error=pedantic -Wno-error=format-security"
```
As mentioned in another comment, I found a way to fix the "-Wno-error=format-security" one so that should be fine.

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp (hunk 9 above)
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp]: I meant that the details were unimplemented. I fixed the message in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/80b6c20eef552e1a3b5176c1bebc5c43da95b827

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/switch_rule/map_based_rule_core.cpp (hunk 10 above)
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/switch_rule/map_based_rule_core.cpp]: In my environment, it does not cause the compilation error. :thinking:  But I fixed it in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/80b6c20eef552e1a3b5176c1bebc5c43da95b827

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp (hunk 8 above)
Comment (KYabuuchi) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp]: As you suggest, it is not necessary to always subscribe to all topics.  :+1: 
I modified it in https://github.com/autowarefoundation/autoware.universe/pull/5846/commits/f358d0d00b8dc0b78093895e6fa45df5945da7c8

With this commit, a subscriber is created only when the corresponding pose_estimator is running. 
(It also contains a minor bug fix. never mind)

Comment (TaikiYamada4) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp]: Thank you for your modification!
I confirmed that no extra topics were subscribed when they are not called for.

File: localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp (hunk 9 above)
Comment (TaikiYamada4) [reply localization/pose_estimator_arbiter/src/pose_estimator_arbiter/pose_estimator_arbiter_core.cpp]: Thank you for your prompt response!
Seems good to me. :+1:

File: localization/pose_estimator_arbiter/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,53 @@
+cmake_minimum_required(VERSION 3.14)
+project(pose_estimator_arbiter)
+
+find_package(autoware_cmake REQUIRED)
+autoware_package()
+
+find_package(glog REQUIRED)
+
+find_package(PCL REQUIRED COMPONENTS common)
+include_directories(SYSTEM ${PCL_INCLUDE_DIRS})
```
[the remark this answers is not in the document]
Comment (TaikiYamada4) [reply localization/pose_estimator_arbiter/CMakeLists.txt:10]: I retried my compilation and it worked with no errors and warnings :+1: 
Thank you for the nice work!

**Comment by KYabuuchi** [comment]
I start [this modification](https://github.com/autowarefoundation/autoware.universe/pull/5846#issuecomment-1891298716), so I revert this PR back to draft once.

File: localization/yabloc/yabloc_image_processing/launch/yabloc_image_processing.launch.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/yabloc/yabloc_image_processing/launch/yabloc_image_processing.launch.xml]: Now, yabloc's modification is completely splited.

**Pull request closed without merging**